// react-stage/src/index.styles.js
export default (state) => {
  const { settings, currentOffset, slideCount, dragging, draggedX } = state;
  const trackWidth = slideCount * 100;
  const slideWidth = 100 / slideCount / settings.slidesToShow;
  const dotWidth = 12;
  const dotPadding = 10;
  return {
    container: {
      position: 'relative',
      width: '100%',
      marginBottom: settings.dots ? `${dotWidth + (dotPadding * 2)}px` : '0px',
      overflow: 'visible'
    },
    stage: {
      position: 'relative',
      width: '100%',
      overflow: 'hidden'
    },
    track: {
      width: `${trackWidth}%`,
      transform: `translateX(calc(${-(currentOffset) * slideWidth}% + ${draggedX}px))`,
      transition: dragging ? 'none' : `transform ${settings.speed / 1000}s ease`
    },
    slide: {
      base: {
        position: 'relative',
        display: 'inline-block',
        width: `${slideWidth}%`,
        zIndex: '1'
      },
      active: {},
    },
    arrow: {
      base: {
        position: 'absolute',
        top: '50%',
        padding: '10px',
        transform: 'translateY(-50%)',
        textTransform: 'uppercase',
        color: '#FFFFFF',
        background: '#000000',
        border: 'none',
        cursor: 'pointer',
        zIndex: '2'
      },
      prev: {
        left: '0px'
      },
      next: {
        right: '0px'
      }
    },
    dot: {
      container: {
        position: 'absolute',
        bottom: `-${dotWidth + (dotPadding * 2)}px`,
        width: '100%',
        textAlign: 'center',
        fontSize: '0px',
        zIndex: '2'
      },
      list: {
        display: 'inline-block',
        margin: '0px',
        padding: '0px'
      },
      item: {
        display: 'inline-block'
      },
      button: {
        display: 'block',
        width: `${dotWidth + (dotPadding * 2)}px`,
        height: `${dotWidth + (dotPadding * 2)}px`,
        margin: '0px',
        padding: `${dotPadding}px`,
        lineHeight: `${dotWidth}px`,
        background: 'none',
        border: 'none',
        cursor: 'pointer'
      },
      circle: {
        base: {
          display: 'block',
          width: `${dotWidth}px`,
          height: `${dotWidth}px`,
          fontSize: '0px',
          background: '#000000',
          border: 'none',
          borderRadius: '50%',
          opacity: '0.5',
          transition: `opacity ${settings.speed / 1000}s ease`
        },
        active: {
          opacity: '1'
        }
      }
    }
  };
};
