// react-stage/demo/components/pages/App.js
import React, { Component } from 'react';
import Helmet from 'react-helmet';
import ReactStage from 'react-stage';
import AppStyles from './AppStyles';
import Arrow from '../elements/arrow';

class App extends Component {
  constructor() {
    super();
    this.state = {
      slider: this.basicSlider()
    };
    this.basicSlider = this.basicSlider.bind(this);
    this.multipleSlider = this.multipleSlider.bind(this);
    this.autoplaySlider = this.autoplaySlider.bind(this);
    this.customArrowTextSlider = this.customArrowTextSlider.bind(this);
    this.customArrowComponentSlider = this.customArrowComponentSlider.bind(this);
    this.externalLibrarySlider = this.externalLibrarySlider.bind(this);
  }

  setSlider(type, callback) {
    let state;
    switch (type) {
      case 'basic':
        state = { slider: this.basicSlider() };
        break;
      case 'multiple':
        state = { slider: this.multipleSlider() };
        break;
      case 'autoplay':
        state = { slider: this.autoplaySlider() };
        break;
      case 'customArrowText':
        state = { slider: this.customArrowTextSlider() };
        break;
      case 'customArrowComponent':
        state = { slider: this.customArrowComponentSlider() };
        break;
      case 'externalLibrary':
        state = { slider: this.externalLibrarySlider() };
        break;
      default:
        break;
    }

    this.setState(state, callback);
  }

  basicSlider() {
    return {
      title: 'Basic Slider',
      markup: (
        <ReactStage>
          <div style={AppStyles.slide}>
            <button>Slide 1</button>
          </div>
          <div style={AppStyles.slide}>
            <button>Slide 2</button>
          </div>
          <div style={AppStyles.slide}>
            <button>Slide 3</button>
          </div>
          <div style={AppStyles.slide}>
            <button>Slide 4</button>
          </div>
          <div style={AppStyles.slide}>
            <button>Slide 5</button>
          </div>
        </ReactStage>
      )
    }
  }

  multipleSlider() {
    return {
      title: 'Multiple Slider',
      markup: (
        <ReactStage
          slidesToShow={2}
          slidesToScroll={2}
        >
          <div style={AppStyles.slide}>
            <span>Slide 1</span>
          </div>
          <div style={AppStyles.slide}>
            <span>Slide 2</span>
          </div>
          <div style={AppStyles.slide}>
            <span>Slide 3</span>
          </div>
          <div style={AppStyles.slide}>
            <span>Slide 4</span>
          </div>
          <div style={AppStyles.slide}>
            <span>Slide 5</span>
          </div>
        </ReactStage>
      )
    };
  }

  autoplaySlider() {
    return {
      title: 'Autoplay Slider',
      markup: (
        <ReactStage
          autoplay={true}
        >
          <div style={AppStyles.slide}>
            <span>Slide 1</span>
          </div>
          <div style={AppStyles.slide}>
            <span>Slide 2</span>
          </div>
          <div style={AppStyles.slide}>
            <span>Slide 3</span>
          </div>
          <div style={AppStyles.slide}>
            <span>Slide 4</span>
          </div>
          <div style={AppStyles.slide}>
            <span>Slide 5</span>
          </div>
        </ReactStage>
      )
    };
  }

  customArrowTextSlider() {
    return {
      title: 'Custom Arrow Text Slider',
      markup: (
        <ReactStage
          arrowPrev="Custom Prev"
          arrowNext="Custom Next"
        >
          <div style={AppStyles.slide}>
            <button>Slide 1</button>
          </div>
          <div style={AppStyles.slide}>
            <button>Slide 2</button>
          </div>
          <div style={AppStyles.slide}>
            <button>Slide 3</button>
          </div>
          <div style={AppStyles.slide}>
            <button>Slide 4</button>
          </div>
          <div style={AppStyles.slide}>
            <button>Slide 5</button>
          </div>
        </ReactStage>
      )
    }
  }

  customArrowComponentSlider() {
    return {
      title: 'Custom Arrow Component Slider',
      markup: (
        <ReactStage
          arrowPrev={(props) => (<Arrow {...props}>Prev Arrow</Arrow>)}
          arrowNext={(props) => (<Arrow {...props}>Next Arrow</Arrow>)}
        >
          <div style={AppStyles.slide}>
            <button>Slide 1</button>
          </div>
          <div style={AppStyles.slide}>
            <button>Slide 2</button>
          </div>
          <div style={AppStyles.slide}>
            <button>Slide 3</button>
          </div>
          <div style={AppStyles.slide}>
            <button>Slide 4</button>
          </div>
          <div style={AppStyles.slide}>
            <button>Slide 5</button>
          </div>
        </ReactStage>
      )
    }
  }

  externalLibrarySlider() {
    return {
      title: 'External Library Slider',
      markup: (
        <div dangerouslySetInnerHTML={{ __html: `
          <div class="react-stage">
            <div style="display: block; text-align: center; padding-top: 100px; padding-bottom: 100px; font-weight: bold; color: rgb(255, 255, 255); background: rgb(201, 0, 0);">
              <button>Slide 1</button>
            </div>
            <div style="display: block; text-align: center; padding-top: 100px; padding-bottom: 100px; font-weight: bold; color: rgb(255, 255, 255); background: rgb(201, 0, 0);">
              <button>Slide 2</button>
            </div>
            <div style="display: block; text-align: center; padding-top: 100px; padding-bottom: 100px; font-weight: bold; color: rgb(255, 255, 255); background: rgb(201, 0, 0);">
              <button>Slide 3</button>
            </div>
            <div style="display: block; text-align: center; padding-top: 100px; padding-bottom: 100px; font-weight: bold; color: rgb(255, 255, 255); background: rgb(201, 0, 0);">
              <button>Slide 4</button>
            </div>
            <div style="display: block; text-align: center; padding-top: 100px; padding-bottom: 100px; font-weight: bold; color: rgb(255, 255, 255); background: rgb(201, 0, 0);">
              <button>Slide 5</button>
            </div>
          </div>
        `}} />
      )
    }
  }

  render() {
    const { slider } = this.state;
    return (
      <div>
        <Helmet>
          <title>React Stage</title>
        </Helmet>
        <h1>React Stage</h1>
        <button onClick={() => this.setSlider('basic')}>Basic Slider</button>
        <button onClick={() => this.setSlider('multiple')}>Multiple Slider</button>
        <button onClick={() => this.setSlider('autoplay')}>Autoplay Slider</button>
        <button onClick={() => this.setSlider('customArrowText')}>Custom Arrow Text Slider</button>
        <button onClick={() => this.setSlider('customArrowComponent')}>Custom Arrow Component Slider</button>
        <button onClick={() => this.setSlider('externalLibrary', () => { window.ReactStage.initialize(); })}>External Library Slider</button>
        <h2>{slider.title}</h2>
        {slider.markup}
      </div>
    );
  }
}

export default App;
